### ISSUE TITLE
index refers to wrong page when indexing words outside of the mainmatter

### ISSUE THREAD

**Issue by ttimbers** [rstudio/bookdown#1274]
I initially opened [a thread about this in the RStudio Community forum](https://community.rstudio.com/t/index-refers-to-wrong-page-when-indexing-words-outside-of-the-mainmatter/118516/2), but after further investigation, I think this might be a bug.

When using bookdown to render to pdf following the instructions in *Write a Book with bookdown and Publish with Chapman & Hall*, we have found (in two separate projects) that items that are indexed in the frontmatter are given the wrong page number - they get pointed to a page of the list of figures, but not to the preface.

So, including something like this in `index.Rmd` (which comes before the `/mainmatter` command):

```
Basic familiarity with version control\index{version control} and ...
```

Gets a page number of "x" - which is one of the pages of the list of figures. I have looked in this `bookdown` repo and see the same thing happens with the demo book:

There is an Index entry in the front matter (`index.Rmd`) for Git Book on [line 54](https://github.com/rstudio/bookdown/blob/6ae8900f1438a463a6e14ab8ffbe75eecb1d59a4/inst/examples/index.Rmd#L54) that looks like this:

```
... and imported the GitBook\index{GitBook} style ...
```

Then in [the pdf](https://bookdown.org/yihui/bookdown/bookdown.pdf) the index entry for Git Book is "ix", which is a list of figure page: 

<img width="555" alt="Screen Shot 2021-10-20 at 8 02 25 PM" src="https://user-images.githubusercontent.com/6865016/138204456-901abfa4-2149-4740-b802-74a3404786eb.png">

<img width="552" alt="Screen Shot 2021-10-20 at 8 02 56 PM" src="https://user-images.githubusercontent.com/6865016/138204471-32512596-ef5a-49db-8ffa-c46cf890c64a.png">

**Comment by ttimbers** [comment]
Just a gentle ping about this issue...

**Comment by cderv** [comment]
Hi @ttimbers, 

Thanks for the ping. That is not an easy one to investigate as it is tied to how latex work. in fact, `\index` feature  is not part of **bookdown** - it is brought by LaTeX by adding 
````latex
  \usepackage{makeidx}
  \makeindex
````
in preamble using header includes and 
```latex
  \printindex
```
where to place the index, usually in after body includes. 

I tested with `bookdown-demo` book and I can"t seem to reproduce this issue. I saw it on the bookdown book too and I wonder if this is not something related to the special class to use for Chapman and Hall: [`krantz.cls` ](https://github.com/rstudio/bookdown/blob/6ae8900f1438a463a6e14ab8ffbe75eecb1d59a4/inst/examples/krantz.cls)

Are you using this documentclass too ? It seems you do  https://github.com/UBC-DSCI/introduction-to-datascience/blob/master/pdf/krantz.cls

I don't really know where to look in **bookdown** for this. I wonder if someone maintaining the krants class could help us see if this maybe be from the class processing or not  🤔 

Did you encounter this issue with another documentclass ?

**Comment by yihui** [comment]
For problems related to `krantz.cls`, usually there is only one person who can help---`Shashi Kumar <texhelp@taylorandfrancis.com>` (Chapman & Hall's TeX help desk).

**Comment by ttimbers** [comment]
Thanks @cderv and @yihui - I will reach out to Shashi!

**Comment by trevorcampbell** [comment]
@cderv and @yihui -- @ttimbers and I worked a bit with Shashi and we found that:
- if we manually compile `bookdown.tex` (output by bookdown) only once, then run `makeindex`, then compile again, the issue we noticed shows up -- the index entries in the preface / frontmatter don't point to the right pages
- if we manually compile `bookdown.tex` twice, then run `makeindex`, then compile again, the index entries all look correct

So basically it looks like this boils down to how many times the latex is being compiled before running `makeindex`. 

Is this something that bookdown controls when it calls `pandoc` (maybe some config params that it passes in?), or is it a bug in `pandoc` where it isn't properly recognizing how many times it needs to compile, or something else?

**Comment by trevorcampbell** [comment]
Ah, I think I may have found the issue:

https://github.com/rstudio/rmarkdown/blob/f43f59ec0578d967d05e46bc7d8bb76e7bfd64a8/R/util.R#L343

looks like it only runs `run_engine()` (which I assume is `pdflatex` or something of that sort) once before `makeindex`. It should run it twice.

I'll test this out on our book, one moment...

**Comment by trevorcampbell** [comment]
Yup, that fixed it. I added a line after line 226 of `latex.R` in tinytex with an extra call to `run_engine()`, and our index entries look good now.

@yihui @cderv -- not sure if that's a reasonable fix or if there's something more clever to do, but at least we've nailed down the problem and a candidate fix. Please let us know how to proceed.

**Comment by cderv** [comment]
Wow thanks a lot for the investigation ! 

Does `makeindex` requires 2 render before in general case ? Or is this specific to our `.tex` file ? I don't know enough LaTeX to know this. Sorry it makes no sense. 

@yihui is this something you would look for a fix in **tinytex** or specific to **bookdown** ? 

It seems that it could happens with any tex file that requires `makeindex` 🤔

**Comment by yihui** [comment]
> @yihui is this something you would look for a fix in **tinytex** or specific to **bookdown** ?

We should fix it in **tinytex**.

**Comment by trevorcampbell** [comment]
> Does makeindex requires 2 render before in general case

Yes, I believe so (it's similar to hyperlinks in the main text)

> It seems that it could happens with any tex file that requires makeindex thinking I wonder if there is something in the log file that would help know than latexmk() needs to still run once more.

Looking at the code right after the `run_engine()` call I pointed out, it seems there is already an if statement to check whether an `.idx` file was created. You could just put a call to `run_engine()` at the start of the if scope.

```
  # some problems only trigger warnings but not errors, e.g.,
  # https://github.com/yihui/tinytex/issues/311 fix them and re-run engine
  if (install_packages && check_extra(logfile)) run_engine()

  # generate index
  if (file.exists(idx <- aux_files['idx'])) {
    idx_engine = getOption('tinytex.makeindex', 'makeindex')
    install_cmd(idx_engine)
    system2_quiet(idx_engine, c(getOption('tinytex.makeindex.args'), shQuote(idx)), error = {
      stop("Failed to build the index via ", idx_engine, call. = FALSE)
    })
  }
```

> I believe that if you set option tinytex.latexmk.emulation to FALSE 

@cderv how do I do this? I assume I have to add something to the `index.Rmd` file?

**Comment by yihui** [comment]
I just pushed the fix. You can test the development version of **tinytex** via

```r
remotes::install_github('yihui/tinytex')
```

Thanks!
